rust/build_assert.rs

```diff
@@ -0,0 +1,50 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! Build-time assertion.
+//!
+//! In Rust it is very common to use assertions extensively and defensively.
+//! Sometimes we know a condition could be checked statically, but it could not
+//! be enforced in Rust (e.g. perform some checks in const functions, but those
+//! functions could still be called in the runtime).
+//!
+//! This crate provides a method `build_error`, which will panic in
+//! compile-time if executed in const context, and will cause a build error
+//! if not executed in compile time and optimizer does not optimise away the
+//! call.
+
+#![no_std]
+#![feature(const_panic, core_panic)]
+
+/// Panics if executed in const context, or triggers a build error if not.
+#[inline(never)]
+#[no_mangle]
+#[track_caller]
+pub const fn build_error(msg: &'static str) {
+    // Could also be panic!(msg) to avoid using unstable feature `core_panic`,
+    // but it is not allowed in Rust 2021, while panic!("{}", msg) could not
+    // yet be used in const context.
+    core::panicking::panic(msg);
+}
+
+/// Asserts that a boolean expression is `true` at compile time.
+///
+/// This will invoke the [`build_error`] function if the compiler or optimizer
+/// cannot guarantee the condition will be evaluated to `true`.
```

Comment:
We should explain that `static_assert!` should be preferred (explaining in which cases it cannot be used). Examples are likely best.

I am not sure the `build_error` explanation is needed (see point above).